clippy_utils/src/ty/mod.rs

```diff
@@ -1376,3 +1376,62 @@ pub fn option_arg_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> Option<Ty<'t
         _ => None,
     }
 }
+
+/// Check if a Ty<'_> of `Iterator` has side effects when iterated over by checking if it
+/// captures any mutable references or equivalents.
+pub fn is_iter_with_side_effects<'tcx>(cx: &LateContext<'tcx>, iter_ty: Ty<'tcx>) -> bool {
+    let Some(iter_trait) = cx.tcx.lang_items().iterator_trait() else {
+        return false;
+    };
+
+    is_iter_with_side_effects_impl(cx, iter_ty, iter_trait)
+}
+
+fn is_iter_with_side_effects_impl<'tcx>(cx: &LateContext<'tcx>, iter_ty: Ty<'tcx>, iter_trait: DefId) -> bool {
+    if let ty::Adt(adt_def, args) = iter_ty.kind() {
+        return adt_def
+            .all_fields()
+            .any(|field| is_ty_with_side_effects(cx, field.ty(cx.tcx, args), iter_trait));
+    }
+
+    false
```

Comment:
Returning false on non-ADTs is not correct. `&mut T` implements `Iterator` and should return true.